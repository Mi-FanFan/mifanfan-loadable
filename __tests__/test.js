const React = require('react')
const renderer = require('react-test-renderer')
const Loadable = require('../src').default

const MyComponent = {
  default(props) {
    return <div>MyComponent {JSON.stringify(props)}</div>;
  }
}

function waitFor(delay) {
  return new Promise(resolve => {
    setTimeout(resolve, delay);
  });
}

function MyLoadingComponent(props) {
  return <div>MyLoadingComponent {JSON.stringify(props)}</div>;
}

function createLoader(delay, loader, error) {
  return () => {
    return waitFor(delay).then(() => {
      if (loader) {
        return loader();
      } else {
        throw error;
      }
    });
  };
}

test('loading success', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, () => MyComponent),
    loading: MyLoadingComponent
  });

  let component1 = renderer.create(<LoadableMyComponent prop="foo" />);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(300);
  expect(component1.toJSON()).toMatchSnapshot(); // loaded

});



test('loading error', async () => {
  let LoadableMyComponent = Loadable({
    loader: createLoader(400, undefined, 3),
    loading: MyLoadingComponent
  });

  let component1 = renderer.create(<LoadableMyComponent prop="foo" />);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(300);
  expect(component1.toJSON()).toMatchSnapshot(); // loaded

});



