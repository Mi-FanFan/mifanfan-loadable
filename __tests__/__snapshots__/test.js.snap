// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading error 1`] = `
<div>
  MyLoadingComponent 
  {"error":null,"timeOut":false}
</div>
`;

exports[`loading error 2`] = `
<div>
  MyLoadingComponent 
  {"error":null,"timeOut":false}
</div>
`;

exports[`loading error 3`] = `
<div>
  MyLoadingComponent 
  {"error":3,"timeOut":false}
</div>
`;

exports[`loading success 1`] = `
<div>
  MyLoadingComponent 
  {"error":null,"timeOut":false}
</div>
`;

exports[`loading success 2`] = `
<div>
  MyLoadingComponent 
  {"error":null,"timeOut":false}
</div>
`;

exports[`loading success 3`] = `
<div>
  MyComponent 
  {"prop":"foo"}
</div>
`;
